src/qibo/parameter.py

```diff
@@ -0,0 +1,141 @@
+        described in the code example above.
+        trainable (list or np.ndarray): array with initial trainable parameters theta
+        features (list or np.ndarray): array containing possible input features x
+        nofeatures (bool): flag to explicitly ban the updating of the features. This simplifies the task of updating Parameter objects simultaneously when some have embedded features and some do not.
+    """
+
+    def __init__(self, func, trainable=None, features=None):
+        self._trainable = trainable
+        self._features = features
+
+        if self.nfeat + self.nparams != func.__code__.co_argcount:
+            raise_error(
+                TypeError,
+                f"{self.nfeat + self.nparams} parameters are provided, but {func.__code__.co_argcount} are required, please initialize features and trainable according to the defined function.",
+            )
+        # lambda function
+        self.lambdaf = func
+
+        # calculate derivatives
+        # maybe here use JAX ?
+        self.derivatives = self._calculate_derivatives()
+
+    def __call__(self):
+        """Update values with trainable parameter and calculate current gate parameter"""
+        return self._apply_func(self.lambdaf)
+
+    @property
+    def nparams(self):
+        """Returns the number of trainable parameters"""
+        try:
+            return len(self._trainable)
+        except TypeError:
+            return 0
+
+    @property
+    def nfeat(self):
+        """Returns the number of features"""
+        return len(self._features) if isinstance(self._features, list) else 0
+
+    @property
+    def ncomponents(self):
+        """Returns the number of elements which compose the Parameter"""
+        return self.nparams + self.nfeat
+
+    def _apply_func(self, function, fixed_params=None):
+        """Applies lambda function and returns final gate parameter"""
+        params = []
+
+        if self._trainable is None:
+            parameter_count = function.__code__.co_argcount
+            params = [0.0] * parameter_count
+
+        else:
+            if self._features is not None:
+                params.extend(self._features)
+            if fixed_params is not None:
+                params.extend(fixed_params)
+            else:
+                params.extend(self._trainable)
+
+        # run function
+        return float(function(*params))
+
+    def _calculate_derivatives(self):
+        """Calculates derivatives w.r.t to all trainable parameters"""
+        vars = []
+        for i in range(self.lambdaf.__code__.co_argcount):
+            vars.append(sp.Symbol(f"p{i}"))
+
+        expr = sp.sympify(self.lambdaf(*vars))
+
+        derivatives = []
+        for i in range(len(vars)):
+            derivative_expr = sp.diff(expr, vars[i])
+            derivatives.append(sp.lambdify(vars, derivative_expr))
+
+        return derivatives
```

Comment:
Since the only thing you're using of `self` is `self.lambdaf`, in this case I would also move this function out of the class, and just accept a single argument (i.e. `lambdaf`, for which you could also use a more standard name like `func`).

Reply:
Exactly!